gobblin-compaction/src/main/java/gobblin/compaction/source/CompactionSource.java

```diff
@@ -84,6 +91,135 @@
+
+      initJobDir(state);
+      copyJarDependencies(state);
+      List<CompactionVerifier> verifiers = suite.getDatasetsFinderVerifiers();
+      DatasetsFinder finder = DatasetUtils.instantiateDatasetFinder(state.getProperties(),
+              getSourceFileSystem(state),
+              DefaultFileSystemGlobFinder.class.getName());
+
+      List<Dataset> datasets = finder.findDatasets();
+      CompactionDatasetIterator iterator = new CompactionDatasetIterator(datasets, verifiers);
+      service.submit(iterator.getDatasetProcessor());
+      return new BasicWorkUnitStream.Builder (iterator).build();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+
+  /**
+   * Iterator that provides {@link WorkUnit}s for all verified {@link Dataset}s
+   */
+  public class CompactionDatasetIterator implements Iterator<WorkUnit> {
+    private LinkedBlockingDeque<WorkUnit> workUnits;
+
+    private List<Dataset> datasets;
+    private List<CompactionVerifier> verifiers;
+    private WorkUnit last;
+    private volatile int unProcessed;
+    private volatile int success;
+
+    /**
+     * Constructor
+     */
+    public CompactionDatasetIterator (List<Dataset> datasets, List<CompactionVerifier> verifiers) {
+      this.datasets = datasets;
+      this.workUnits = new LinkedBlockingDeque<>();
+      this.verifiers = verifiers;
+      this.unProcessed = datasets.size();
+      this.success = 0;
+      this.last = null;
+    }
+
+    /**
+     * Check if any {@link WorkUnit} is available. The producer is {@link CompactionDatasetIterator#getDatasetProcessor()}
+     * @return true when a new {@link WorkUnit}  is available
+     *         false when producer exits
+     */
+    public boolean hasNext () {
+      try {
+        while (true) {
+          if (last != null) {
+            log.debug ("hasNext() is true because cache is not empty");
+            return true;
+          }
+          if (this.unProcessed > 0) {
+            last = this.workUnits.poll(1, TimeUnit.SECONDS);
+            if (last == null)
+              log.debug ("Waiting for producer to complete...");
+          } else {
+            if (workUnits.size() > 0) {
+              log.debug ("hasNext() has new element");
+              return true;
+            }
+            log.debug ("hasNext() returns false because producer is complete");
+            return false;
+          }
+        }
+      } catch (InterruptedException e) {
+        log.error(e.toString());
+        return false;
+      }
+    }
+
+    /**
+     * Obtain next available {@link WorkUnit}.
+     * Block the consumer thread until a new {@link WorkUnit} is provided. Otherwise throw an exception
+     */
+    public WorkUnit next () {
+      if (hasNext()) {
+        if (last != null) {
+          WorkUnit tmp = last;
+          last = null;
+          log.debug ("next() pops out a workunit");
+          return tmp;
+        }
```

Comment:
you should probably have `else { throw new IllegalStateException()} ` to catch the case where `hasNext()` returns true but `last` doesn't get populated.